drivers/dac/ad9144/ad9144.c

```diff
@@ -363,6 +364,72 @@ static int32_t ad9144_pll_setup(struct ad9144_dev *dev,
 	return ret;
 }
 
+int32_t ad9144_set_nco(struct ad9144_dev *dev, int32_t f_carrier_khz,
+		       int16_t phase)
+{
+	uint32_t modulation_type, phase_offset;
+	bool sel_sideband = false;
+	uint64_t ftw;
+	uint8_t reg;
+	int32_t ret;
+	uint8_t i;
+
+	if (phase > abs(180))
+		return -1;
+	if (f_carrier_khz < 0) {
+		f_carrier_khz *= -1;
+		sel_sideband = true;
+	}
+
+	if (f_carrier_khz >= dev->sample_rate_khz / 2) {
+		modulation_type = MODULATION_TYPE(0);
+	} else if (dev->sample_rate_khz == f_carrier_khz * 4) {
+		modulation_type = MODULATION_TYPE(2);
+	} else if (dev->sample_rate_khz == f_carrier_khz * 8) {
+		modulation_type = MODULATION_TYPE(2);
+	} else {
+		modulation_type = MODULATION_TYPE(1);
+	}
+	ret = ad9144_spi_read(dev, REG_DATAPATH_CTRL, &reg);
+	if (ret < 0)
+		return ret;
+
+	reg = (reg & ~MODULATION_TYPE_MASK) | modulation_type;
+	if (sel_sideband)
+		reg |= SEL_SIDEBAND;
+	else
+		reg &= ~SEL_SIDEBAND;
+
+	ret = ad9144_spi_write(dev, REG_DATAPATH_CTRL, reg);
+	if (ret < 0)
+		return ret;
+
+	ftw = ((1ULL << 48) / dev->sample_rate_khz * f_carrier_khz);
+	for (i = 0; i < 6; i++) {
+		ret = ad9144_spi_write(dev, REG_FTW0 + i,
+				       (ftw >> (8 * i)) & 0xFF);
+		if (ret < 0)
+			return ret;
+	}
+
+	phase_offset = (phase/180) * (1 << 15);
+	ret = ad9144_spi_write(dev, REG_NCO_PHASE_OFFSET0, phase_offset & 0xFF);
+	if (ret < 0)
+		return ret;
+	ret = ad9144_spi_write(dev, REG_NCO_PHASE_OFFSET1, (phase_offset >> 8) &
+			       0xFF);
+	if (ret < 0)
+		return ret;
+
+	if (modulation_type  == MODULATION_TYPE(1)) {
```

Comment:
maybe add macro for each modulation type?

Reply:
Added some comments